code/game/machinery/doors/firedoor.dm

```diff
@@ -188,30 +188,52 @@
 		for(var/obj/machinery/firealarm/fire_panel in place.firealarms)
 			fire_panel.set_status()
 
+/obj/machinery/door/firedoor/proc/CalculateWatchedTurfs()
+	watched_turfs = list()
+	for(var/dir in GLOB.cardinals)
+		var/turf/checked_turf = get_step(get_turf(src),dir)
+		watched_turfs |= checked_turf
+		RegisterSignal(checked_turf, COMSIG_TURF_EXPOSE, .proc/check_atmos)
+	watched_turfs |= get_turf(src)
+
 /obj/machinery/door/firedoor/proc/check_atmos(datum/source)
+	SIGNAL_HANDLER
+	var/found_turf = FALSE
+
 	if(!COOLDOWN_FINISHED(src, detect_cooldown))
 		return
-	if(alarm_type)
+	if(alarm_type == FIRELOCK_ALARM_TYPE_GENERIC)
 		return
+	
 	for(var/area/place in affecting_areas)
 		if(!place.fire_detect) //if any area is set to disable detection
 			return
 
-	var/turf/my_turf = source
-	if(!my_turf)
-		my_turf = get_turf(src)
+	if(!watched_turfs)
+		watched_turfs = CalculateWatchedTurfs()
+
+	for(var/turf/checked_turf in watched_turfs)
+		if(source == checked_turf)
+			found_turf = TRUE
```

Comment:
Why do we do this exactly?

Reply:
sanity checking to make sure the registered signals corresponds to a good tile, It wouldnt work without it somehow

Reply:
This shouldn't be needed, signals straying like that is a sign that something underlying is borked

Reply:
This is still true